### PR TITLE
Bump scenes to 0.17.2

### PR DESCRIPTION
**Pull request by dprokop**
Closes https://github.com/grafana/grafana/pull/70505

Renovate bot failed to update scenes in https://github.com/grafana/grafana/pull/70505. The reason is something that I didn't take into consideration. It's a consequence of panel options and field config types being part of @grafana/schema package (https://github.com/grafana/grafana/pull/64723). In scenes, the new API PanelBuilders (ref https://github.com/grafana/scenes/pull/225) is using the types available via @grafana/schema in order to provide support when building scene panels. However, those types are not exported from schema package as a public API. Instead we are relying on absolute import paths of those types i.e.:

```
import {
  Options as BarGaugeOptions,
  defaultOptions as defaultBarGaugeOptions,
} from '@grafana/schema/dist/esm/raw/composable/bargauge/panelcfg/x/BarGaugePanelCfg_types.gen';
```

We want all common types for panels to be exported under unified name (`Options` and `FieldConfig`) - see discussions in https://github.com/grafana/grafana/pull/64723, hence the import paths.

But - this won't work in core Grafana repo, as this path is invalid - there's basically no `dist` version available in the core repo, we rely on workspace packages. Hence the failure in https://github.com/grafana/grafana/pull/70505

In order to get latest scenes workin in core repo, we need to slighlty modify jest, typescript and webpack config, so that those absolute imports are resolved from the workspace @grafana/schema. This PR is bringing those config updates as well as API updates in core where scenes are used.

### CODE REVIEW

**Comment by dprokop** [comment]
@ashharrison90 @jackw  - pinging you here since you are way more fluent with those configs than I am :)

**Review by ashharrison90 — CHANGES_REQUESTED** [review]
**tl;dr: i don't know exactly what the solution is, but i'm 99% sure it should be made in `@grafana/schema` and not in the consumers.**

so i've been reading the linked issue, still not 100% up to speed, and i'm also by no means an expert at module exports etc

i don't think we should be doing anything in `grafana/grafana`. this feels like the responsibility of `@grafana/schema` to export things properly. either normally via an index file at the root, or if we're absolutely wedded to the path imports then at the very least a consumer should be able to do 
```tsx
import { whatever } from '@grafana/schema/raw/composable/bargauge/panelcfg/x/BarGaugePanelCfg_types.gen';`
```
and the schema package routes you to the correct thing automatically.

hypothetically imagine we decided to expose commonjs imports as well as esm. then doing:
```tsx
const { whatever } = require('@grafana/schema/raw/composable/bargauge/panelcfg/x/BarGaugePanelCfg_types.gen');
```
should take us to a different folder (probably `/dist/commonjs/...`). but we wouldn't ask the consumer to specify a different path, the package should just handle that.

**Comment by jackw** [comment]
>this feels like the responsibility of `@grafana/schema` to export things properly. either normally via an index file at the root, or if we're absolutely wedded to the path imports then at the very least a consumer should be able to do
> 
> ```tsx
> import { whatever } from '@grafana/schema/raw/composable/bargauge/panelcfg/x/BarGaugePanelCfg_types.gen';`
> ```
> 
> and the schema package routes you to the correct thing automatically.
> 
> hypothetically imagine we decided to expose commonjs imports as well as esm. then doing:
> 
> ```tsx
> const { whatever } = require('@grafana/schema/raw/composable/bargauge/panelcfg/x/BarGaugePanelCfg_types.gen');
> ```
> 
> should take us to a different folder (probably `/dist/commonjs/...`). but we wouldn't ask the consumer to specify a different path, the package would handle that properly.

I believe what you're describing is technically possible with the [exports property](https://nodejs.org/docs/latest-v18.x/api/packages.html#conditional-exports) in package.json.

Replacing the `main` property in `grafana-schema/package.json` for something like the following should allow these paths to resolve correctly from within grafana/grafana without the need to touch any configs.

```json
  "exports": {
    ".": {
      "types": "./src/index.ts",
      "import": "./src/index.ts",
      "require": "./src/index.ts"
    },
    "./src/raw/*": "./src/raw/*.ts",
    "./dist/esm/raw/*": "./src/raw/*.ts"
  },
```
However we'll need to supply a `publishConfig.exports` property to change them when packaging the tarball for NPM so external consumers (e.g. plugins / scenes) can continue to access the exports as they currently do.

**Comment by dprokop** [comment]
@jackw / @ashharrison90  - thanks a ton for review! Let's see how changing schema config will help this, will push an update soon.

**Comment by dprokop** [comment]
@ashharrison90 - I'm not sure you had a chance to have a look at this, but I spent some time trying to apply suggestions from @jackw. In order to get it working we need to change the tsconfig to `moduleResolution: node16` (ref https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#ecmascript-module-support-in-nodejs). This has plenty of implications that are well described in the TS docs. 

Feels like a bigger project to migrate to ES modules and allow the suggested config. I'm not sure it's worth the effort as of now, I can't see clear benefits for core  of the change. 

I see an alternative path - that would make the paths equivalent between the dist and workspace schema, it's basically generating the structure of the same path in the dist grafana/schema, that would be exactly the same as the one in the workspace. So the published grafana/schema would have the following structure:

```
- dist/
- src/ //here comes the problematic types imported via full path
- ...
```

I'm not convinced for this approach as we would have to change the way we build, bundle, and publish grafana/schema. Feels like it would be a way more distributed set of changes to get that done compared to the approach I'm proposing with this PR.

WDYT?

**Comment by jackw** [comment]
> Feels like a bigger project to migrate to ES modules and allow the suggested config. I'm not sure it's worth the effort as of now, I can't see clear benefits for core of the change.

I agree that if the proposed changes cause a lot of work across the codebase we should handle it in a separate PR.

>I see an alternative path - that would make the paths equivalent between the dist and workspace schema, it's basically generating the structure of the same path in the dist grafana/schema, that would be exactly the same as the one in the workspace. So the published grafana/schema would have the following structure:

Whilst it's good that alternatives have been researched I think for now the proposed changes in this PR are probably the best way forward.

**Review by jackw — APPROVED** [review]
LGTM! 🚀

**Comment by ashharrison90** [comment]
@dominik sorry for the delay, looking at this now 👀 

what should we be checking to make sure things are working? if i undo the 3 changes to webpack/jest/tsconfig and apply jack's change to the `@grafana/schema` package.json the tests in grafana now run correctly. it doesn't look like i even need to set `moduleResolution: node16`. i'm probably missing something obvious 😄 

even if we do need to set `moduleResolution`, wouldn't we only need to do that in the `@grafana/schema` package? 🤔

**Comment by jackw** [comment]
> it doesn't look like i even need to set moduleResolution: node16. i'm probably missing something obvious

It will fix the issues with the tests and webpack should also resolve things correctly but it introduces issues with type checking. Typescript won't resolve the `exports` without moving the project over to use `nodenext` for module resolution and most likely will also require adding `type: module` to the root package.json.

> even if we do need to set moduleResolution, wouldn't we only need to do that in the @grafana/schema package? 🤔

It seems the consumer needs to set it to [tell typescript](https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing) to use the exports field.

**Comment by dprokop** [comment]
@ashharrison90  - as @jackw writes. It's all about typecheck. Just try `yarn start` and you'll see issues :)

> It seems the consumer needs to set it to [tell typescript](https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing) to use the exports field.

Precisely. And this leads to loads of build errors.

**Review by ashharrison90 — APPROVED** [review]
ah, it was something obvious 😄 

thanks for bearing with me @dprokop! idk how much input you have with the `@grafana/schema` package, but i'd honestly question if the value we get from these path imports outweighs having to write mappings for all our dev tooling in `grafana/grafana`... it **will** bite us at some point 😞 

we can go with this for now 👍

**Comment by dprokop** [comment]
>  idk how much input you have with the @grafana/schema package, but i'd honestly question if the value we get from these path imports outweighs having to write mappings for all our dev tooling in grafana/grafana

@ashharrison90 - well, the value is simple, you can see it in this docs: https://grafana.github.io/scenes/docs/visualizations#standard-grafana-visualizations. We either have a typed API, or we don't. From the standpoint of API ease of use, I think a few lines of additional config doesn't hurt - and let's consider this config as a temporary thing. I'm happy to collab on getting the core repo to support node16 moduleResolution. 

And thanks again @ashharrison90 and @jackw!
